packages/protocol/contracts/L1/libs/LibAuction.sol

```diff
@@ -0,0 +1,190 @@
+            revert L1_MAX_FEE_PER_GAS_EXCEEDED(
+                config.maxFeePerGasForAuctionBid
+            );
+        }
+
+        // deposit is the auction bidders "how sure they are they will be able to prove the block"
+        // a higher deposit ensures they are quite confident.
+
+        if (!isAuctionOpen(config, state, batchId)) {
+            revert L1_AUCTION_CLOSED_FOR_BATCH();
+        }
+
+        uint256 weight = calculateBidWeight(deposit, minFeePerGas);
+        uint256 existingBidWeight = state.blockAuctionBids[batchId].weight;
+
+        if (weight < existingBidWeight) {
+            revert L1_INSUFFICIENT_WEIGHT({
+                bidWeight: weight,
+                existingBidWeight: existingBidWeight
+            });
+        }
+
+        TaikoData.Bid memory bid = state.blockAuctionBids[batchId];
+
+        TaikoToken tkoToken = TaikoToken(
+            resolver.resolve("taiko_token", false)
+        );
+
+        if (bid.account != address(0)) {
+            // refund deposit if there is an existing bid
+            try tkoToken.transfer(bid.account, bid.deposit) {} catch {
+                // allow to fail in case they have a bad onTokenReceived
+            }
+        }
+
+        if (!tkoToken.transferFrom(msg.sender, address(this), deposit)) {
+            revert L1_TRANSFER_FROM_FAILED();
+        }
+
+        uint256 auctionStartedAt = bid.auctionStartedAt > 0
+            ? bid.auctionStartedAt
+            : block.timestamp;
+
+        state.blockAuctionBids[batchId] = TaikoData.Bid({
+            account: msg.sender,
+            feePerGas: minFeePerGas,
+            deposit: deposit,
+            weight: weight,
+            auctionStartedAt: auctionStartedAt,
+            batchId: batchId
+        });
+
+        emit AuctionBid({
+            batchId: batchId,
+            bidder: msg.sender,
+            bidAt: block.timestamp,
+            minFeePerGas: minFeePerGas,
+            deposit: deposit,
+            weight: weight,
+            auctionStartedAt: auctionStartedAt
+        });
+    }
+
+    // TODO: this is just placeholder weighting rn
+    function calculateBidWeight(
+        uint256 deposit,
+        uint256 minFeePerGas
+    ) internal pure returns (uint256) {
+        uint256 depositWeight = 1e2;
+        uint256 minFeePerGasWeight = 1e18;
+        return
+            (deposit * depositWeight) +
+            (minFeePerGas * minFeePerGasWeight) /
+            10000;
+    }
+
+    function isAuctionOpen(
+        TaikoData.Config memory config,
+        TaikoData.State storage state,
+        uint256 batchId
+    ) internal view returns (bool) {
+        // auction is always open if there is no winner.
+        if (state.blockAuctionBids[batchId].account == address(0)) return true;
+
+        // if auctionLengthInSeconds hasnt passed, the auction is still open
+        if (
+            state.blockAuctionBids[batchId].auctionStartedAt +
```

Comment:
If we use the following check for aunctions:

```
uint firstBatch = (nextBlockId << 8) + 1;
uint lastBatch = firstBatch + config.auctionBatchGap
require(batchId >=firstBatch && batchId <= lastBatch);
```

Nobody can win a batch the (in the same L1 tx) prove the blocks immediately, which create outbid opportunities for other bidders. If the delay between bidding and proving should be larger, then we can change the above to:


```
uint firstBatch = (nextBlockId << 8) + 2; // changed from 1 to 2.
uint lastBatch = firstBatch + config.auctionBatchGap
require(batchId >=firstBatch && batchId <= lastBatch);
```

Now maybe we can get rid of auctionLengthInSeconds.